### PR TITLE
Refactoring of TupleGen and EnumGen

### LINKED ISSUE #458: [CodeGen] toString of Enums
I would like to specialise the `toString` method of enums for to specific use cases.

- As a rule, the unit value should never be printed. For example, if there is a `case Red` then it it is implicitly `case Red(Unit)` but should still be printed as just `Red`. The only exception is when printing the unit value itself which should be printed as `()`. 

- The `Cons` tag of the `List` enum should be printed infix as `::`. Hence the list `Cons(1, Cons(2, Cons(3, Nil)))` should be printed as `1 :: 2 :: 3 :: Nil`.

### LINKED ISSUE #459: [CodeGen] toString of Tuples
The `toString` of tuples should simple be `(elm1, ...m elmn)`. 

E.g. you could have `(42, 21)` which should print as `(42, 21)`.

### LINKED ISSUE #460: [CodeGen] Tags
The code for a generated tag looks approximately like:

```
public final class Cons implements EnumInterface {
    public Object value;
    public String tag;

    public Cons(Object var1) {
        this.value = var1;
        this.tag = "Cons";
    }
}

```

It would be better if the `Object value` was `private` and if `String tag` was not a field, but simply the value returned by `getTag`.

### LINKED ISSUE #461: [CodeGen] Tuples
A current tuple looks like:

```
public final class Tuple implements TupleInterface {
    public int field0;
    public int field1;

    public Tuple(int var1, int var2) {
        this.field0 = var1;
        this.field1 = var2;
    }
```

We should refactor this such that the fields are hidden behind getters/setters. 

This will likely help with some other things I have planned down the road.

### PR DESCRIPTION
**Pull request by FlaminGuy**
Changes in this pull request:
1- Adding WrappedTypes to the backend to replace the some/None. Make the logic more clear
2- Fixing `toString()` of tuple. This fixes #459 
3- Replacing `Unit` by a singleton object #449 
4- Fixing `toString()` of enums. This fixes #458 
5- Privatizing `value` field and deleting `tag` field on the enum. Fixes #460
6- Adding getters and setters to tuples. Fixes #461 
7- Cases with `Unit` field now follow a singleton model. #463
8- Changed singleton equality testing for enums.

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/api/UnitClass.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.api;
+
+/**
+ * UnitClass which is used to represent `Unit`
+ */
+public class UnitClass {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/UnitClass.java]: I would prefer if we just use the names:

- `UnitClass -> Unit`
- `TagInterface -> Enum`
- `TupleInterface -> Tuple`.

The interface/class thing is not really idiomatic Java, Scala, or Flix.

File: main/src/ca/uwaterloo/flix/api/UnitClass.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    public String toString() {
+        return "()";
+    }
+
+    /**
+     * Returns true if `obj` is instance of `UnitClass'
+     */
+    @Override
+    public boolean equals(Object obj) {
+        return obj instanceof UnitClass;
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/UnitClass.java]: `return this eq that`

File: main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala [hunk 3]
```diff
@@ -100,13 +100,14 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root]{
     // 3. Create Enum Type info
     val allEnums : List[(Type, (String, Type))] = root.definitions.values.flatMap(x => CodegenHelper.findEnumCases(x.exp)).toList
 
-    val enumTypeInfo: Map[(Type, String), QualName] = allEnums.map{ case (tpe, (name, subType)) =>
+    val enumTypeInfo: Map[(Type, String), (QualName, Boolean)] = allEnums.map{ case (tpe, (name, subType)) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: What is this boolean?

The type signature of this map is a bit hard to read; it goes from a pair of (Type, String) to (QualName, Boolean) - what are these components?

EDIT: I guess the boolean is whether it is singleton.

File: main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala [hunk 4]
```diff
@@ -528,63 +530,67 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root]{
...
-      visitor.visitTypeInsn(NEW, decorate(clazzName))
-      visitor.visitInsn(DUP)
-      // Evaluating the single argument of the class constructor
-      compileExpression(prefix, functions, declarations, interfaces, enums, visitor, entryPoint)(exp)
-      // Calling the constructor of the class
-      visitor.visitMethodInsn(INVOKESPECIAL, decorate(clazzName), "<init>", s"(${desc})V", false)
-
+      val (clazzName, isSingleton) = enums(tpe, tag)
+
+      if(isSingleton) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: Add comments to explain.

File: main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala [hunk 5]
```diff
@@ -528,63 +530,67 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root]{
...
-        case Type.Apply(Type.FTuple(_), lst) => TupleClassName(lst.map(transformTypeToOptionType))
+        case Type.Apply(Type.FTuple(_), lst) => TupleClassName(lst.map(typeToWrappedType))
         case _ => throw InternalCompilerException(s"Unexpected type: `${base.tpe}`")
       }
       // evaluating the `base`
       compileExpression(prefix, functions, declarations, interfaces, enums, visitor, entryPoint)(base)
-      // Fetching the field containing the element specified by the `offset`
-      visitor.visitFieldInsn(GETFIELD, decorate(clazzName), s"field$offset",desc)
+      // Invoking `getField${offset}()` method for fetching the field
+      visitor.visitMethodInsn(INVOKEVIRTUAL, decorate(clazzName), s"getField$offset", s"()$desc", false)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: Could we name these `getIndex` and `setIndex`?

File: main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala [hunk 6]
```diff
@@ -42,13 +41,50 @@ object CodegenHelper {
   val JavaVersion = V1_8
 
   /**
+    * Wrapper around similar types
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala]: Explain purpose. Maybe with an example.

File: main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala [hunk 7]
```diff
@@ -261,17 +283,61 @@ object EnumGen extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
   }
 
   /**
+    * Initializing `getInstance` static field if the `value` field can be `UnitClass`
+    * @param visitor class visitor
+    * @param className Qualified name of the class
+    */
+  private def initializeStaticField(visitor: ClassWriter, className: EnumClassName) = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala]: Better name? Like `loadUnitValue` or similar?

File: main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala (hunk 3 above)
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: I am wondering if passing the `ExecutableAst.Case` ast node would not simplify things?

Then you could have a helper method `def isSingleton(caze: ExecutableAst.Case)` basically trying to reuse the AST ast much as possible. I understand that for the specialized types you need to carry more information than the enum/case, but having these nodes available might still simplify things.

Comment (FlaminGuy) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: You are right, this is for whether it is boolean or not.

Comment (FlaminGuy) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: `ExecutableAst.Case` doesn't help that much here except for finding if it is a singleton or not.
This map is used to find the right class that implements the enum and we need type of the field to do it yet case doesn't provide this info.
Best thing I can do is to change it to `Map[(Type, String), (QualName, Boolean)]`. Do you prefer this?

Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: Yes, that would be better. 

You might also find this informative: https://existentialtype.wordpress.com/2011/03/15/boolean-blindness/

Comment (FlaminGuy) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: Interesting article. Also interesting blog.

**Comment by FlaminGuy** [comment]
I fixed the issues.
Only thing is, changing `UnitClass` to `Unit` ended up having huge conflicts with `scala.Unit`. My build was working fine but travis was failing because my build wasn't building the whole thing from scratch.

**Comment by magnus-madsen** [comment]
Nice!

**Pull request merged**